src/analysis/processing/qgsalgorithmpointstopaths.cpp

```diff
@@ -0,0 +1,334 @@
+  QgsFeature f;
+  const double totalPoints = source->featureCount() > 0 ? 100.0 / source->featureCount() : 0;
+  long currentPoint = 0;
+  feedback->setProgressText( QObject::tr( "Loading points…" ) );
+  while ( fit.nextFeature( f ) )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+    feedback->setProgress( currentPoint * totalPoints );
+
+    if ( f.hasGeometry() && ! f.geometry().isNull() )
+    {
+      expressionContext.setFeature( f );
+      const QVariant orderValue = orderExpression.evaluate( &expressionContext );
+      const QVariant groupValue = groupExpressionString.isEmpty() ? QVariant() : groupExpression.evaluate( &expressionContext );
+
+      if ( ! allPoints.keys().contains( groupValue ) )
+        allPoints[ groupValue ] = QVector< QPair< QVariant, const QgsPoint * > >();
+      const QgsPoint *point = qgsgeometry_cast< const QgsPoint * >( f.geometry().constGet()->clone() );
+      allPoints[ groupValue ] << QPair< QVariant, const QgsPoint *>( orderValue, point );
+    }
+    ++currentPoint;
+  }
+
+  int pathCount = 0;
+  currentPoint = 0;
+  QHashIterator< QVariant, QVector< QPair< QVariant, const QgsPoint * > > > hit( allPoints );
+  feedback->setProgressText( QObject::tr( "Creating paths…" ) );
+  while ( hit.hasNext() )
+  {
+    hit.next();
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+    auto pairs = hit.value();
+
+    if ( naturalSort )
+    {
+      std::sort( pairs.begin(),
+                 pairs.end(),
+                 [&collator]( const QPair< const QVariant, const QgsPoint * > &pair1,
+                              const QPair< const QVariant, const QgsPoint * > &pair2 )
+      {
+        return collator.compare( pair1.first.toString(), pair2.first.toString() ) < 0;
+      } );
+    }
+    else
+    {
+      std::sort( pairs.begin(),
+                 pairs.end(),
+                 []( const QPair< const QVariant, const QgsPoint * > &pair1,
+                     const QPair< const QVariant, const QgsPoint * > &pair2 )
+      {
+        return qgsVariantLessThan( pair1.first, pair2.first );
+      } );
+    }
+
+
+    QVector<QgsPoint> pathPoints;
+    for ( QVector< QPair< QVariant, const QgsPoint * > >::ConstIterator pit = pairs.constBegin(); pit != pairs.constEnd(); ++pit )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+      feedback->setProgress( currentPoint * totalPoints );
+      pathPoints.append( *pit->second );
+      ++currentPoint;
+    }
+    if ( pathPoints.size() < 2 )
+    {
+      feedback->pushInfo( QObject::tr( "Skipping path with group %1 : insufficient vertices" ).arg( hit.key().toString() ) );
+      continue;
+    }
+    if ( closePaths && pathPoints.size() > 2 && pathPoints.first() != pathPoints.last() )
+      pathPoints.append( pathPoints.first() );
+
+    QgsFeature outputFeature;
+    QgsAttributes attrs;
+    if ( ! groupExpressionString.isEmpty() )
+      attrs.append( hit.key() );
```

Comment:
else append a QVariant(), or the attributes will go out-of-sync with the fields

Reply:
Fields are handled similarly (group field is skipped if unneeded, this is how the python version works):
https://github.com/qgis/QGIS/blob/1afd4048e985c81320b3ce3c15bdb64920fcc523/src/analysis/processing/qgsalgorithmpointstopaths.cpp#L183-L191

Reply:
Ah, missed that -- thanks!